crates/core/src/ring.rs

```diff
@@ -431,12 +566,240 @@ impl Ring {
+                } else {
+                    check_interval.tick().await;
+                    continue;
+                }
+            }
+
+            let mut should_swap = {
+                let peers = self.connections_by_location.read();
+                should_swap(
+                    peers
+                        .values()
+                        .flatten()
+                        .filter(|conn| {
+                            conn.open_at.elapsed() > CONNECTION_AGE_THRESOLD
+                                && !live_tx_tracker.has_live_connection(&conn.location.peer)
+                        })
+                        .map(|conn| &conn.location),
+                )
+            };
+            if !should_swap.is_empty() {
+                self.fast_acquisition
+                    .store(false, std::sync::atomic::Ordering::Release);
+                let ideal_location = {
+                    self.topology_manager
+                        .write()
+                        .get_best_candidate_location()
+                        .expect("we only swap when we have some established neighbors already")
+                };
+                self.acquire_new(
+                    ideal_location,
+                    &missing.values().collect::<ArrayVec<_, { 5120 / 80 }>>(),
+                    &notifier,
+                )
+                .await
+                .map_err(|error| {
+                    tracing::warn!(?error, "shutting down connection maintenance task");
+                    error
+                })?;
+                for peer in should_swap.drain(..) {
+                    notifier
+                        .send(Either::Right(crate::message::NodeEvent::DropConnection(
+                            peer,
+                        )))
+                        .await
+                        .map_err(|error| {
+                            tracing::debug!(?error, "shutting down connection maintenance task");
+                            error
+                        })?;
+                }
+            }
+            check_interval.tick().await;
+        }
+    }
+
+    #[tracing::instrument(level = "debug", skip(self, notifier))]
+    async fn acquire_new(
+        &self,
+        ideal_location: Location,
+        skip_list: &[&PeerKey],
+        notifier: &EventLoopNotificationsSender,
+    ) -> Result<(), DynError> {
+        let msg = {
+            let conns = &*self.connections_by_location.read();
+            let Some(msg) = Self::find_optimal_query_target(
+                (self.own_location(), ideal_location),
+                conns
+                    .iter()
+                    .flat_map(|(loc, conns)| conns.iter().map(|conn| (*loc, &conn.location))),
+                skip_list,
+            ) else {
+                return Ok(());
+            };
+            msg
+        };
+        notifier.send(Either::Left((msg.into(), None))).await?;
+        Ok(())
+    }
+
+    fn find_optimal_query_target<'a>(
```

Comment:
I'm unclear on the purpose of this, isn't this what the Router is for?

Reply:
I guess we could be using the router here yes. 

The goal is to send a request to join peer C from peer A to an optimal query target, peer B. Here the optimal was the closes to the ideal location which was an heuristic that was simple enough (after all, the target which is the closest to the ideal location, is likely to have the most connections to peers closest to the ideal location. Peer B here is just acting as a mere proxy to forward that request to the final receiver of the request so not sure if using the router would be necessary but I guess it won't hurt.